### PR TITLE
Example dataset usage in downstream machine learning modeling

### PR DESCRIPTION
**Pull request by michaelmaser**
Notebook example reproducing a yield regression model of the Perera dataset, pulling from ord-data.

PTAL and let me know if you have any suggestions!

(This will automatically become available on the binder repo, is that right?)

### CODE REVIEW

**Comment by skearnes** [comment]
Looks like the notebook test is failing?

File: examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb [hunk 1]
```diff
@@ -0,0 +1,2515 @@
...
+    "from glob import glob"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Download data from ord-data pull request and load to Dataset message\n",
```
Comment (skearnes) [inline examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb:52]: My preference would be to have one cell that downloads the file when it is executed and a second cell for loading it into a dataset. Since many people will likely use this notebook as a template, I'd prefer to avoid the extra complexity of checking for existing files or alternative suffixes.

File: examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb [hunk 2]
```diff
@@ -0,0 +1,2515 @@
...
+     "execution_count": 4,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# Ensure dataset validates\n",
+    "valid_output = validations.validate_message(data)\n",
+    "\n",
+    "# Check errors\n",
```
Comment (skearnes) [inline examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb:87]: Since validate_message by default raises an exception on error, I think this cell could be simplified to just `validations.validate_message(data)`.

File: examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb [hunk 3]
```diff
@@ -0,0 +1,2515 @@
...
+    "# Example application of ORD datasets for downstream machine learning modeling"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Reproducing neural network predictor of reaction yields from Granda et al. Nature 2018, 559, 377, DOI: 10.1038/s41586-018-0307-8.<br>\n",
+    "Data from Perera et al. Science 2018, 359, 429, DOI: 10.1126/science.aap9112.<br>\n",
+    "ORD dataset location: https://github.com/open-reaction-database/ord-data/tree/main/data/68<br>\n",
```
Comment (skearnes) [inline examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb:16]: Please put the full URL here; there could be other datasets with the 68 prefix.

File: examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb [hunk 4]
```diff
@@ -0,0 +1,2515 @@
...
+    "          'solvent']\n",
+    "\n",
+    "# Create one-hot encoded input dataset\n",
+    "ohe_df = pd.get_dummies(df[input_cols], prefix=prefix)\n",
+    "\n",
+    "# Add yield column to ohe dataset\n",
+    "ohe_df['yield'] = df[model_cols[-1]]/100\n",
+    "\n",
+    "# Shuffle dataset\n",
+    "ohe_df = ohe_df.sample(frac=1)\n",
```
Comment (skearnes) [inline examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb:1603]: consider adding a random seed here for reproducible output

File: examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb [hunk 5]
```diff
@@ -0,0 +1,2515 @@
...
+      "(3456, 35) (3456,)\n",
+      "(576, 35) (576,)\n",
+      "(1728, 35) (1728,)\n",
+      "True\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Set training and test set parameters\n",
+    "train_size = math.floor(0.6 * len(inputs))\n",
```
Comment (skearnes) [inline examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb:1656]: consider using sklearn.model_selection.train_test_split (twice) instead

File: examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb [hunk 6]
```diff
@@ -0,0 +1,2515 @@
...
+    "\n",
+    "# Setup callback to load weights from best epoch later on\n",
+    "# Saving model weights from epoch with lowest validation loss\n",
+    "checkpoint_folder = os.path.join(os.getcwd(), 'saved_model')\n",
+    "if not os.path.exists(checkpoint_folder):\n",
+    "    os.makedirs(checkpoint_folder)\n",
+    "checkpoint_filepath = os.path.join(checkpoint_folder, 'saved_model')\n",
+    "model_checkpoint_callback = tf.keras.callbacks.ModelCheckpoint(\n",
+    "    filepath=checkpoint_filepath,\n",
+    "    save_weights_only=True,\n",
```
Comment (skearnes) [inline examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb:2354]: Why save_weights_only?

File: examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb [hunk 7]
```diff
@@ -0,0 +1,2515 @@
...
+    "ax2.plot(y_test, preds, 'bo')\n",
+    "ax2.plot([0, 1], 'r')\n",
+    "ax2.set_ylabel('predicted yield')\n",
+    "ax2.set_xlabel('ground truth yield')\n",
+    "ax2.set_title('predictions vs ground truth')\n",
+    "ax2.text(0.6, 0, f'test_RMSE: {round(rmse, 3)}')\n",
+    "ax2.text(0.6, -0.05, f'test_R2: {round(r2, 3)}')\n",
+    "\n",
+    "# Save and view performance plots\n",
+    "plt.savefig('ORD_Perera_HTE_yield_regression_MLP_performance.png', dpi=100)\n",
```
Comment (skearnes) [inline examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb:2449]: Any reason to save these figures?

File: examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb [hunk 8]
```diff
@@ -0,0 +1,2515 @@
...
+    "import numpy as np\n",
+    "import tensorflow as tf\n",
+    "import matplotlib.pyplot as plt\n",
+    "import os\n",
+    "import wget\n",
+    "\n",
+    "from rdkit import Chem\n",
+    "from rdkit.Chem import AllChem\n",
+    "from sklearn.model_selection import train_test_split\n",
+    "from sklearn.metrics import r2_score\n",
```
Comment (skearnes) [inline examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb:42]: import modules, not functions; https://google.github.io/styleguide/pyguide.html#22-imports

**Review by skearnes — CHANGES_REQUESTED** [review]
Looks great Mike, thanks!

File: examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb (hunk 1 above)
Comment (michaelmaser) [reply examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb:52]: Got it, thanks!

File: examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb (hunk 2 above)
Comment (michaelmaser) [reply examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb:87]: Done, thanks!

File: examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb (hunk 3 above)
Comment (michaelmaser) [reply examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb:16]: Done, thanks!

File: examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb (hunk 5 above)
Comment (michaelmaser) [reply examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb:1656]: Done, thanks!

File: examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb (hunk 6 above)
Comment (michaelmaser) [reply examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb:2354]: Fixed, thanks! Copy, pasted from other notebook.

File: examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb (hunk 7 above)
Comment (michaelmaser) [reply examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb:2449]: Nope. Fixed, thanks!

File: examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb (hunk 8 above)
Comment (michaelmaser) [reply examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb:42]: Fixed, thanks!

**Review by skearnes — APPROVED** [review]

File: examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb [hunk 9]
```diff
@@ -0,0 +1,2609 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Download dataset from ord-data\n",
+    "url = 'https://github.com/open-reaction-database/ord-data/blob/main/data/68/ord_dataset-68cb8b4b2b384e3d85b5b1efae58b203.pb.gz?raw=true'\n",
```
Comment (brilee) [inline examples/applications/Perera_Science_Granda_Nature_Suzuki/Granda_Perera_ml_example.ipynb:52]: How do you construct this URL for an arbitrary dataset? Is there a web UI where we could get that link?
